### DISCUSSION TITLE
Every changes triggers onDidFilesChange

### DISCUSSION THREAD

**Discussion by cenkce** [eclipse-theia/theia#9548] [Q&A]
Every file changes in a workspace triggers onDidFilesChange that I never watch. Is it correct behavior? How can I handle it?
- Theia 1.4.0
- Macos 10.15.7

**Comment by vince-fugnitto** [comment]
@cenkce thank you for the discussion, can you clarify by providing a use-case and why you expect `onDidFilesChange` to not trigger?

>  that I never watch

I assume you are explicitly watching for certain files, but there is also the general file-watching mechanism which will fire the event whenever there is a change in the workspace, for example if a file is opened and modified, or if a file is modified outside the application.

**Reply by cenkce** [reply]
Hi, there are so many services that modify different folders so that onDidFilesChange is triggered by those. This creates some performance issue in order to continuously filter files changes. Is there any way to prevent that?

**Reply by cenkce** [reply]
And I comment out 'watch' line in the code and changes keep coming. In that case why I use watch method?

**Reply by vince-fugnitto** [reply]
> there any services that modify different folders so that onDidFilesChange is triggered by those. This creates some performance issue in order to continuously filter files changes. Is there any way to prevent that?

Unfortunately I am not aware of any mechanism to filter out unwanted changes from the file-watchers ([axosoft/nsfw](https://github.com/Axosoft/nsfw)), perhaps filtering may work but it might be specific to different operating system. What types of services modify changes in bulk which cause you performance issues?

> And I comment out 'watch' line in the code and changes keep coming. In that case why I use watch method?

There is a global watcher on the workspace root which notifies of changes in the actual workspace. One may use `watch` to watch files outside the workspace such as preferences (`~/.theia/settings.json`), configs and/or other files.

**Reply by cenkce** [reply]
Is there a watch or exclude option or similar in a settings.json or can I receive and update configuration of the global watch?

**Reply by cenkce** [reply]
I found that "files.watcherExclude" in the settings.json so can I programmatically modify that?

**Reply by vince-fugnitto** [reply]
> I found that "files.watcherExclude" in the settings.json so can I programmatically modify that?

@cenkce that might be a good enough solution for your use-case, you have a few options:
- modify the value as a user (this will not apply to other users, and might not be the best solution overall for your application)
- you can modify the default value by referencing the preference as part of your application's `package.json` - essentially we allow application developers the ability to modify the default value of any preference through the `package.json` like [so](https://github.com/theia-ide/theia-apps/blob/d329db260cc8e96759241198153d9d3fd731f32e/theia-cpp-docker/latest.package.json#L10-L12).
- you can make the modifications in your application through a [custom extension](https://theia-ide.org/docs/authoring_extensions) which will either modify the default or use the `PreferenceService` to set the initial value to something you require.

**Reply by cenkce** [reply]
Yes I use custom extension. PreferenceService looks the best way. Thanks.

**Reply by cenkce** [reply]
@vince-fugnitto files.watcherExclude has no effect. I try it inside of workflow/.theia and ~/.theia but never change.

```
{
    "workbench.iconTheme": "theia-file-icons",
    "files.watcherExclude": {
        "**/.git/objects/**": true,
        "**/.git/subtree-cache/**": true,
        "**/node_modules/**": true,
        "**/dist/**": true,
        "**/scripts/generated/**": true
    },
    "editor.autoSave": "off",
    "application.confirmExit": "always"
}
```

**Reply by cenkce** [reply]
and also this one `"editor.autoSave": "off",` never worked

**Reply by vince-fugnitto** [reply]
@cenkce preferences should be either:
- user scope: `~/.theia/settings.json`
- workspace scope: `{workspace-path}/.theia/settings.json`
- folder scope

Note that preferences with a more precise scope will have priority.

Can you confirm where the settings you mentioned in  https://github.com/eclipse-theia/theia/discussions/9548#discussioncomment-826649 exist?

**Reply by cenkce** [reply]
Yes they exist but no effect. I can't understand what's wrong.

**Reply by vince-fugnitto** [reply]
> Yes they exist but no effect. I can't understand what's wrong.

I have never seen this error before, perhaps try setting these values through the UI:
- <kbd>F1</kbd>
- search and confirm the value `open preferences`
- search for the corresponding preference in the view and modify the value
